agent/tests/BrowserFake.php

```diff
@@ -34,13 +49,21 @@ public function post(string $url, array $headers, string $body): PromiseInterfac
         $response = array_shift($this->pendingResponses);
 
         if ($response === null) {
-            throw new RuntimeException('A request was made but there are no more responses.');
+            throw new RuntimeException('A request was made but there are no more responses: ['.json_encode([
+                'url' => $url,
+            ], flags: JSON_THROW_ON_ERROR).']');
         }
 
-        if (is_array($response)) {
-            return reject(new $response[0]($response[1]));
-        }
+        $this->processingResponses[] = $response;
+
+        return $response->toPromise()->finally(function () use ($response) {
+            $index = array_search($response, $this->processingResponses, true);
+
+            if ($index === false) {
+                throw new RuntimeException('Was unable to find the processing response. Something is wrong.');
+            }
 
-        return $response->toPromise();
+            unset($this->processingResponses[$index]);
+        });
```

Comment:
This allows us to simulate a request taking time.

When a request is sent we now mark it as processing. Once the response as been received it is then cleared from the processing list.